src/BSON/Value.c

```diff
@@ -0,0 +1,631 @@
+PHONGO_VALUE_TYPE_ISSER(isDouble, BSON_TYPE_DOUBLE);
+PHONGO_VALUE_TYPE_ISSER(isDocument, BSON_TYPE_DOCUMENT);
+PHONGO_VALUE_TYPE_ISSER(isInt32, BSON_TYPE_INT32);
+PHONGO_VALUE_TYPE_ISSER(isInt64, BSON_TYPE_INT64);
+PHONGO_VALUE_TYPE_ISSER(isMaxKey, BSON_TYPE_MAXKEY);
+PHONGO_VALUE_TYPE_ISSER(isMinKey, BSON_TYPE_MINKEY);
+PHONGO_VALUE_TYPE_ISSER(isNull, BSON_TYPE_NULL);
+PHONGO_VALUE_TYPE_ISSER(isObjectId, BSON_TYPE_OID);
+PHONGO_VALUE_TYPE_ISSER(isRegex, BSON_TYPE_REGEX);
+PHONGO_VALUE_TYPE_ISSER(isSymbol, BSON_TYPE_SYMBOL);
+PHONGO_VALUE_TYPE_ISSER(isTimestamp, BSON_TYPE_TIMESTAMP);
+PHONGO_VALUE_TYPE_ISSER(isUndefined, BSON_TYPE_UNDEFINED);
+PHONGO_VALUE_TYPE_ISSER(isUTCDateTime, BSON_TYPE_DATE_TIME);
+PHONGO_VALUE_TYPE_ISSER(isUtf8, BSON_TYPE_UTF8);
+
+static PHP_METHOD(MongoDB_BSON_Value, isCode)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	RETURN_BOOL(intern->value.value_type == BSON_TYPE_CODE || intern->value.value_type == BSON_TYPE_CODEWSCOPE);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, isInt)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	RETURN_BOOL(intern->value.value_type == BSON_TYPE_INT32 || intern->value.value_type == BSON_TYPE_INT64);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, isNumber)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	RETURN_BOOL(intern->value.value_type == BSON_TYPE_INT32 || intern->value.value_type == BSON_TYPE_INT64 || intern->value.value_type == BSON_TYPE_DOUBLE);
+}
+
+/* MongoDB\BSON\Value object handlers */
+static zend_object_handlers php_phongo_handler_value;
+
+static void php_phongo_value_free_object(zend_object* object)
+{
+	php_phongo_value_t* intern = Z_OBJ_VALUE(object);
+
+	zend_object_std_dtor(&intern->std);
+
+	bson_value_destroy(&intern->value);
+
+	if (intern->properties) {
+		zend_hash_destroy(intern->properties);
+		FREE_HASHTABLE(intern->properties);
+	}
+}
+
+static zend_object* php_phongo_value_create_object(zend_class_entry* class_type)
+{
+	php_phongo_value_t* intern = zend_object_alloc(sizeof(php_phongo_value_t), class_type);
+
+	zend_object_std_init(&intern->std, class_type);
+	object_properties_init(&intern->std, class_type);
+
+	intern->std.handlers = &php_phongo_handler_value;
+
+	return &intern->std;
+}
+
+static zend_object* php_phongo_value_clone_object(phongo_compat_object_handler_type* object)
+{
+	php_phongo_value_t* intern;
+	php_phongo_value_t* new_intern;
+	zend_object*        new_object;
+
+	intern     = Z_OBJ_VALUE(PHONGO_COMPAT_GET_OBJ(object));
+	new_object = php_phongo_value_create_object(PHONGO_COMPAT_GET_OBJ(object)->ce);
+
+	new_intern = Z_OBJ_VALUE(new_object);
+	zend_objects_clone_members(&new_intern->std, &intern->std);
+
+	bson_value_copy(&intern->value, &new_intern->value);
+
+	return new_object;
+}
+
+// TODO: can we even compare values?
```

Comment:
I'm considering implementing a comparison handler that is able to indicate equality, i.e. returns `0` if type and value match, and returns `1` if they don't match. The use case here would be to allow for easier comparison in tests (e.g. for `__clone` or `__serialize`) and potentially for users, as checking equality/inequality is a common use case. I'll note here (and later in the docs) that using `<`, `>`, `<=`, `>=`, and `<=>` would not produce consistent results due to us always returning `1` for unequal objects.

A workaround for the latter would be to order `bson_value_t` by type first, and by value (using `memcmp`) second, i.e.:
```php
return (value1.type <=> value2.type) ?: (value1.value <=> value2.value);
```

While this doesn't handle comparison of equal values of different types (e.g. `int32(2) == int64(2)`), it would provide a more consistent behaviour than the first suggestion by ensuring that `(value1 > value2) === (value2 < value1)`. This would not be the case if we always returned `1` on inequality, as that would mean that `value1 > value2` and `value2 > value1` would both be true at the same time.

Reply:
Oh, I forgot `bson_compare()` was so simplistic. Would you mind moving the above comments to [PHPC-2209](https://jira.mongodb.org/browse/PHPC-2209) so we can continue the conversation there?

Reply:
I expanded the description of PHPC-2209 to capture this.